File: backend/src/main/java/com/amg/digitalitzacio/auth/application/UserService.java
package com.amg.digitalitzacio.auth.application;

import com.amg.digitalitzacio.auth.api.dto.*;
import com.amg.digitalitzacio.auth.domain.*;
import com.amg.digitalitzacio.shared.security.Role;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.UUID;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final TenantRepository tenantRepository;
    private final PasswordEncoder passwordEncoder;

    @Transactional
    public UserResponse createUser(@Valid CreateUserRequest request, Role callerRole) {
        if (userRepository.existsByEmail(request.email())) {
            throw new IllegalArgumentException("Email ja registrat");
        }

        // ADMIN només pot crear CLIENT — forcem el rol
        var targetRole = request.role();
        if (callerRole == Role.ADMIN && targetRole != Role.CLIENT) {
            targetRole = Role.CLIENT;
        }

        if (targetRole == Role.CLIENT && request.tenantId() == null) {
            throw new IllegalArgumentException("El rol CLIENT requereix un tenantId");
        }

        if (request.tenantId() != null && !tenantRepository.existsById(request.tenantId())) {
            throw new IllegalArgumentException("Tenant no trobat");
        }

        var user = User.builder()
                .email(request.email())
                .passwordHash(passwordEncoder.encode(request.password()))
                .name(request.name())
                .role(targetRole)
                .tenantId(targetRole == Role.CLIENT ? request.tenantId() : null)
                .isActive(true)
                .isBlocked(false)
                .failedAttempts(0)
                .build();

        user = userRepository.save(user);
        return toResponse(user);
    }

    public Page<UserResponse> listUsers(Pageable pageable, Role role, UUID tenantId, String search) {
        String searchPattern = (search != null && !search.isBlank()) ? "%" + search.toLowerCase() + "%" : null;
        return userRepository.findAllFiltered(role, tenantId, searchPattern, pageable).map(this::toResponse);
    }

    public UserResponse getUser(UUID id) {
        var user = userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Usuari no trobat"));
        return toResponse(user);
    }

    @Transactional
    public UserResponse updateUser(UUID id, @Valid UpdateUserRequest request, Role callerRole) {
        var user = userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Usuari no trobat"));

        // ADMIN no pot canviar rol ni modificar usuaris que no siguin CLIENT
        if (callerRole == Role.ADMIN) {
            if (user.getRole() != Role.CLIENT) {
                throw new IllegalArgumentException("ADMIN només pot modificar usuaris CLIENT");
            }
            if (request.role() != null && request.role() != user.getRole()) {
                throw new IllegalArgumentException("ADMIN no pot canviar el rol d'un usuari");
            }
            if (request.tenantId() != null) {
                throw new IllegalArgumentException("ADMIN no pot canviar el tenant d'un usuari");
            }
        }

        if (request.email() != null) {
            if (!request.email().equals(user.getEmail()) && userRepository.existsByEmail(request.email())) {
                throw new IllegalArgumentException("Email ja registrat");
            }
            user.setEmail(request.email());
        }
        if (request.name() != null) user.setName(request.name());
        if (request.role() != null && callerRole != Role.ADMIN) user.setRole(request.role());
        if (request.tenantId() != null && callerRole != Role.ADMIN) user.setTenantId(request.tenantId());
        if (request.isActive() != null) user.setIsActive(request.isActive());

        user = userRepository.save(user);
        return toResponse(user);
    }

    @Transactional
    public void deactivateUser(UUID id) {
        var user = userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Usuari no trobat"));
        user.setIsActive(false);
        userRepository.save(user);
    }

    @Transactional
    public void unlockUser(UUID id) {
        var user = userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Usuari no trobat"));
        user.resetFailedAttempts();
        userRepository.save(user);
    }

    private UserResponse toResponse(User user) {
        UserResponse.TenantRef tenantRef = null;
        if (user.getTenantId() != null) {
            var tenant = tenantRepository.findById(user.getTenantId()).orElse(null);
            if (tenant != null) {
                tenantRef = new UserResponse.TenantRef(tenant.getId(), tenant.getName());
            }
        }
        return new UserResponse(
                user.getId(), user.getEmail(), user.getName(), user.getRole(),
                tenantRef, user.getIsActive(), user.getIsBlocked(),
                user.getLastLoginAt(), user.getCreatedAt());
    }
}

File: frontend/src/services/userService.ts
import { apiFetch } from './api';

export interface UserResponse {
  id: string;
  email: string;
  name: string;
  role: 'SUPER_ADMIN' | 'ADMIN' | 'CLIENT';
  tenant: { id: string; name: string } | null;
  isActive: boolean;
  isBlocked: boolean;
  lastLoginAt: string | null;
  createdAt: string;
}

export interface PageResponse<T> {
  content: T[];
  page: number;
  size: number;
  totalElements: number;
  totalPages: number;
}

export interface CreateUserRequest {
  email: string;
  password: string;
  name: string;
  role: 'SUPER_ADMIN' | 'ADMIN' | 'CLIENT';
  tenantId?: string;
}

export interface UpdateUserRequest {
  email?: string;
  name?: string;
  role?: 'SUPER_ADMIN' | 'ADMIN' | 'CLIENT';
  tenantId?: string;
  isActive?: boolean;
}

export interface ListUsersParams {
  page?: number;
  size?: number;
  role?: string;
  tenantId?: string;
  search?: string;
}

export async function listUsers(params: ListUsersParams = {}): Promise<PageResponse<UserResponse>> {
  const searchParams = new URLSearchParams();
  if (params.page !== undefined) searchParams.set('page', String(params.page));
  if (params.size !== undefined) searchParams.set('size', String(params.size));
  if (params.role) searchParams.set('role', params.role);
  if (params.tenantId) searchParams.set('tenantId', params.tenantId);
  if (params.search) searchParams.set('search', params.search);
  const qs = searchParams.toString();
  return apiFetch(`/users${qs ? `?${qs}` : ''}`);
}

export async function getUser(id: string): Promise<UserResponse> {
  return apiFetch(`/users/${id}`);
}

export async function createUser(data: CreateUserRequest): Promise<UserResponse> {
  return apiFetch('/users', {
    method: 'POST',
    body: JSON.stringify(data),
  });
}

export async function updateUser(id: string, data: UpdateUserRequest): Promise<UserResponse> {
  return apiFetch(`/users/${id}`, {
    method: 'PUT',
    body: JSON.stringify(data),
  });
}

export async function deleteUser(id: string): Promise<void> {
  await apiFetch(`/users/${id}`, { method: 'DELETE' });
}

export async function unlockUser(id: string): Promise<{ message: string }> {
  return apiFetch(`/users/${id}/unlock`, { method: 'POST' });
}
